File: Spaces/Models/SpaceMessage.swift
import Foundation

struct MessageReplyContext: Hashable {
    let messageId: String
    let senderName: String
    let type: String
    let preview: String
}

struct SpaceMessage: Identifiable, Hashable {
    let id: String
    let spaceId: String?
    let senderId: String?
    let senderName: String
    let senderEmoji: String?
    let type: MessageType
    let encryptionVersion: String
    let deleted: Bool
    let text: String?
    let media: SpaceMedia?
    let createdAt: Date?
    let updatedAt: Date?
    let timestamp: String
    let isOutgoing: Bool
    let status: String?
    let deliveryStatus: String?
    let isEdited: Bool
    let editedAt: Date?
    let replyContext: MessageReplyContext?
    let reactions: [MessageReaction]

    init(
        id: String,
        spaceId: String? = nil,
        senderId: String? = nil,
        senderName: String,
        senderEmoji: String? = nil,
        type: MessageType = .text,
        encryptionVersion: String = "none",
        deleted: Bool = false,
        text: String? = nil,
        media: SpaceMedia? = nil,
        createdAt: Date? = nil,
        updatedAt: Date? = nil,
        timestamp: String,
        isOutgoing: Bool,
        status: String? = nil,
        deliveryStatus: String?,
        isEdited: Bool = false,
        editedAt: Date? = nil,
        replyContext: MessageReplyContext? = nil,
        reactions: [MessageReaction] = []
    ) {
        self.id = id
        self.spaceId = spaceId
        self.senderId = senderId
        self.senderName = senderName
        self.senderEmoji = senderEmoji
        self.type = type
        self.encryptionVersion = encryptionVersion
        self.deleted = deleted
        self.text = text
        self.media = media
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.timestamp = timestamp
        self.isOutgoing = isOutgoing
        self.status = status
        self.deliveryStatus = deliveryStatus
        self.isEdited = isEdited
        self.editedAt = editedAt
        self.replyContext = replyContext
        self.reactions = reactions
    }

    init(
        id: UUID,
        spaceId: String? = nil,
        senderId: String? = nil,
        senderName: String,
        senderEmoji: String? = nil,
        type: MessageType = .text,
        encryptionVersion: String = "none",
        deleted: Bool = false,
        text: String? = nil,
        media: SpaceMedia? = nil,
        createdAt: Date? = nil,
        updatedAt: Date? = nil,
        timestamp: String,
        isOutgoing: Bool,
        status: String? = nil,
        deliveryStatus: String?,
        isEdited: Bool = false,
        editedAt: Date? = nil,
        replyContext: MessageReplyContext? = nil,
        reactions: [MessageReaction] = []
    ) {
        self.init(
            id: id.uuidString,
            spaceId: spaceId,
            senderId: senderId,
            senderName: senderName,
            senderEmoji: senderEmoji,
            type: type,
            encryptionVersion: encryptionVersion,
            deleted: deleted,
            text: text,
            media: media,
            createdAt: createdAt,
            updatedAt: updatedAt,
            timestamp: timestamp,
            isOutgoing: isOutgoing,
            status: status,
            deliveryStatus: deliveryStatus,
            isEdited: isEdited,
            editedAt: editedAt,
            replyContext: replyContext,
            reactions: reactions
        )
    }
}

File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/data/model/SpaceMessage.kt
package com.arcinteractive.spaces.data.model

import java.util.Date

data class MessageReplyContext(
    val messageId: String,
    val senderName: String,
    val type: String,
    val preview: String
)

data class LinkPreviewData(
    val originalUrl: String,
    val canonicalUrl: String? = null,
    val domain: String,
    val title: String,
    val summary: String? = null,
    val siteName: String? = null,
    val imageDataBase64: String? = null,
    val imageMimeType: String? = null
)

enum class SpaceLinkModuleType(
    val title: String,
    val icon: String,
    val emoji: String
) {
    Announcements("Announcement", "megaphone.fill", "\uD83D\uDCE2"),
    Polls("Poll", "chart.bar.xaxis", "\uD83D\uDCCA"),
    Files("File", "folder.fill", "\uD83D\uDCC1"),
    Events("Event", "calendar", "\uD83D\uDCC5"),
    Rooms("Room", "bubble.left.and.text.bubble.right.fill", "\uD83D\uDCAC"),
    Media("Media", "photo.on.rectangle", "\uD83D\uDDBC\uFE0F"),
    Lists("List", "checklist", "✅"),
    Notes("Note", "note.text", "📝")
}

data class SpaceLinkAttachment(
    val id: String,
    val moduleType: SpaceLinkModuleType,
    val targetId: String,
    val title: String,
    val subtitle: String? = null,
    val icon: String = moduleType.icon,
    val version: Int = 1
) {
    val searchableText: String
        get() = listOf(title, subtitle.orEmpty(), moduleType.title).joinToString("\n")
}

data class SpaceMessage(
    val id: String,
    val spaceId: String? = null,
    val senderId: String? = null,
    val senderName: String,
    val senderEmoji: String? = null,
    val type: MessageType = MessageType.Text,
    val encryptionVersion: String = "none",
    val deleted: Boolean = false,
    val text: String? = null,
    val media: SpaceMedia? = null,
    val mediaItems: List<SpaceMedia> = emptyList(),
    val createdAt: Date? = null,
    val updatedAt: Date? = null,
    val timestamp: String,
    val isOutgoing: Boolean,
    val status: String? = null,
    val deliveryStatus: String? = null,
    val isEdited: Boolean = false,
    val editedAt: Date? = null,
    val replyContext: MessageReplyContext? = null,
    val linkPreview: LinkPreviewData? = null,
    val spaceLinks: List<SpaceLinkAttachment> = emptyList(),
    val reactions: List<MessageReaction> = emptyList(),
    val localDeliveryState: LocalMessageDeliveryState? = null,
    val localFailureMessage: String? = null
) {
    val resolvedMediaItems: List<SpaceMedia>
        get() = if (mediaItems.isNotEmpty()) mediaItems else listOfNotNull(media)

    val primaryMedia: SpaceMedia?
        get() = resolvedMediaItems.firstOrNull()

    val hasMediaAttachments: Boolean
        get() = resolvedMediaItems.isNotEmpty()
}
